### PR TITLE
metrics: Instrument batches store

### PR DESCRIPTION
**Pull request by eseliger**
Adding basic metrics on call counts/durations/errors for different store methods, so we can get some insights into our system.

### CODE REVIEW

File: enterprise/internal/batches/store/batch_changes.go [hunk 1]
```diff
@@ -283,7 +306,12 @@ type GetBatchChangeDiffStatOpts struct {
...
 }
 
-func (s *Store) GetBatchChangeDiffStat(ctx context.Context, opts GetBatchChangeDiffStatOpts) (*diff.Stat, error) {
+func (s *Store) GetBatchChangeDiffStat(ctx context.Context, opts GetBatchChangeDiffStatOpts) (stat *diff.Stat, err error) {
+	ctx, endObservation := s.operations.getBatchChangeDiffStat.With(ctx, &err, observation.Args{LogFields: []log.Field{
+		log.Int("batchChangeID", int(opts.BatchChangeID)),
+	}})
+	defer endObservation(1, observation.Args{})
+
 	authzConds, err := database.AuthzQueryConds(ctx, s.Handle().DB())
```
Comment (courier-new) [inline enterprise/internal/batches/store/batch_changes.go:315]: TIL you can still use `:=` to reassign a variable _if it's in a multi-variable assignment_. Does this still work if _both_ variables were already declared, or only if at least one of them is new?

**Comment by eseliger** [comment]
Some GitHub fatfingering, the changes in here were approved here: https://github.com/sourcegraph/sourcegraph/pull/23486 (the diff was the complete change).

**Review by efritz — COMMENTED** [review]
💦

**Review by courier-new — APPROVED** [review]
Lgtm. 👍 Questions about how this works, though.

1. How does one test that this is working? Are you running `prometheus` locally and looking for a certain output? I'd also love to get the 30-second overview from you or someone else at some point how to find a specific metric we're tracking in prod. Is that generally through the Grafana dashboards?
2. I see the signature of `endObservation` is `type FinishFunc func(count float64, args Args)`, but I don't see a comment as to what either argument is intended for. What is the `count` of, and what are the `args` for? It feels strange that we **always** call it with exactly `1` and the empty `observation.Args{}` struct -- can we not set these as defaults for some reason? I literally [can't find any other occurrences](https://k8s.sgdev.org/search?q=context:global+repo:%5Egithub%5C.com/sourcegraph/sourcegraph%24+endObservation%5C%28%5B2-9%5D&patternType=regexp) of us calling this with a `count` other than `1`, for example:

	image

	Anyway that's not really a you problem, I was mostly just surprised by how chunky a PR to add "basic metrics" to store methods got, and I know go code can get very verbose, but this stood out to me.

**Comment by courier-new** [comment]
Oh haha maybe I should move my questions over to that PR

File: enterprise/internal/batches/store/batch_changes.go (hunk 1 above)
Comment (eseliger) [reply enterprise/internal/batches/store/batch_changes.go:315]: One of them must be new, otherwise the compiler complains 🤷‍♂️

**Comment by eseliger** [comment]
> Oh haha maybe I should move my questions over to that PR

All good, from a diff perspective, this is the right place to post it. I don't know why github showed the entire diff in the other _stacked_ PR, too.

> How does one test that this is working? Are you running prometheus locally and looking for a certain output? I'd also love to get the 30-second overview from you or someone else at some point how to find a specific metric we're tracking in prod. Is that generally through the Grafana dashboards?

Yes! I run `sg run-set monitoring` and it emits both the grafana and prometheus URLs to look at. You can find all metrics we currently have in the grafana boards in the site-admin navigation:

Happy to walk you through this in one of our syncs or in a 1-1 though, if you like.

> I see the signature of endObservation is type FinishFunc func(count float64, args Args), but I don't see a comment as to what either argument is intended for. What is the count of, and what are the args for? It feels strange that we always call it with exactly 1 and the empty observation.Args{} struct -- can we not set these as defaults for some reason? I literally can't find any other occurrences of us calling this with a count other than 1, for example:

It's additional labels and the call count (which for us is always 1, I think the implementation is rather generic but we only use it for the inc by 1 case right now). I agree it's very verbose, but I didn't want to change the entire thing in a 30 minute "lets get this done" run in between meetings :) At this point I just stamp those things off as "it's Go 🤷‍♂️" mentally when I see them 🤦 😆

**Pull request merged**